### PR TITLE
OpticalFlowAdapter generalization

### PR DESCRIPTION
**Pull request by szalpal**

### CODE REVIEW

File: dali/aux/optical_flow/optical_flow_adapter.h [hunk 1]
```diff
@@ -16,14 +16,23 @@
...
-using Backend = kernels::StorageGPU;
+template<typename ComputeBackend>
+struct Compute2Storage {
+  using type = kernels::StorageCPU;
+};
+
+template<>
+struct Compute2Storage<kernels::ComputeGPU> {
+  using type = kernels::StorageGPU;
+};
```
Comment (klecki) [inline dali/aux/optical_flow/optical_flow_adapter.h]:
```
template <typename ComputeBackend>
using compute_to_storage_backend_t = typename compute_to_storage_backend<ComputeBackend>::type;
```

File: dali/aux/optical_flow/optical_flow_adapter.h [hunk 2]
```diff
@@ -16,14 +16,23 @@
...
 namespace detail {
 
-using Backend = kernels::StorageGPU;
+template<typename ComputeBackend>
+struct Compute2Storage {
+  using type = kernels::StorageCPU;
+};
+
+template<>
+struct Compute2Storage<kernels::ComputeGPU> {
```
Comment (klecki) [inline dali/aux/optical_flow/optical_flow_adapter.h]:
```suggestion
struct compute_to_storage_backend<kernels::ComputeGPU> {
```
Traits are traditionally written with lowercase snake case.

File: dali/aux/optical_flow/optical_flow_stub.h [hunk 3]
```diff
@@ -25,18 +26,48 @@ namespace optical_flow {
  * Stub implementation for OpticalFlow.
  * All it does is assign two values: 666.f and 333.f to output_image
  */
-class DLL_PUBLIC OpticalFlowStub : public OpticalFlowAdapter {
+template<typename ComputeBackend>
+class DLL_PUBLIC OpticalFlowStub : public OpticalFlowAdapter<ComputeBackend> {
+  using StorageBackend = typename OpticalFlowAdapter<ComputeBackend>::StorageBackend;
```
Comment (klecki) [inline dali/aux/optical_flow/optical_flow_stub.h]:
```suggestion
  using StorageBackend = compute_to_storage_backend_t<ComputeBackend>;
```

File: dali/aux/optical_flow/optical_flow_adapter.h [hunk 4]
```diff
@@ -39,19 +48,20 @@ struct OpticalFlowParams {
...
 
-template<typename Backend, typename DataType, int ndim>
-using TV = dali::kernels::TensorView<Backend, DataType, ndim>;
-
+using dali::kernels::TensorView;
 
+template<typename ComputeBackend>
 class DLL_PUBLIC OpticalFlowAdapter {
+ protected:
+  using StorageBackend = typename detail::Compute2Storage<ComputeBackend>::type;
```
Comment (klecki) [inline dali/aux/optical_flow/optical_flow_adapter.h]:
```suggestion
  using StorageBackend = compute_to_storage_backend_t<ComputeBackend>;
```

File: dali/aux/optical_flow/optical_flow_adapter_test.cc [hunk 5]
```diff
@@ -13,25 +13,53 @@
 // limitations under the License.
 
 #include <gtest/gtest.h>
+#include <dali/aux/optical_flow/optical_flow_stub.h>
```
Comment (jantonguirao) [inline dali/aux/optical_flow/optical_flow_adapter_test.cc]: use `#include ""` for dali headers

File: dali/aux/optical_flow/optical_flow_adapter_test.cc [hunk 6]
```diff
@@ -13,25 +13,53 @@
...
+  CUDA_CALL(cudaMalloc(reinterpret_cast<void **>(&tvout_data),
+                       kTestDataSize * sizeof(StubValueType)));
+
+  TensorView<StorageGPU, uint8_t, 3> tvref, tvin;
+  TensorView<StorageGPU, float, 3> tvout(tvout_data, {1, 1, 2});
+  OpticalFlowParams params;
+  std::unique_ptr<OpticalFlowAdapter<ComputeGPU>> of(new OpticalFlowStub<ComputeGPU>(params));
+  of->CalcOpticalFlow(tvref, tvin, tvout);
+
+  std::unique_ptr<float> host(new float[kTestDataSize]);
```
Comment (jantonguirao) [inline dali/aux/optical_flow/optical_flow_adapter_test.cc]: this is leaking. Why not using a vector?

File: dali/aux/optical_flow/optical_flow_adapter.h (hunk 2 above)
Comment (szalpal) [reply dali/aux/optical_flow/optical_flow_adapter.h]: Done

File: dali/aux/optical_flow/optical_flow_stub.h [hunk 7]
```diff
@@ -15,8 +15,9 @@
 #ifndef DALI_AUX_OPTICAL_FLOW_OPTICAL_FLOW_STUB_H_
 #define DALI_AUX_OPTICAL_FLOW_OPTICAL_FLOW_STUB_H_
 
-#include "dali/kernels/tensor_view.h"
-#include "dali/aux/optical_flow/optical_flow_adapter.h"
+#include <dali/kernels/tensor_view.h>
+#include <dali/aux/optical_flow/optical_flow_adapter.h>
```
Comment (jantonguirao) [inline dali/aux/optical_flow/optical_flow_stub.h]: `#include ""`

File: dali/aux/optical_flow/optical_flow_stub.h [hunk 8]
```diff
@@ -25,18 +26,48 @@ namespace optical_flow {
...
 
+
+template<>
+inline void OpticalFlowStub<kernels::ComputeCPU>::CalcOpticalFlow(
+        dali::kernels::TensorView<dali::kernels::StorageCPU, const uint8_t, 3> reference_image,
+        dali::kernels::TensorView<dali::kernels::StorageCPU, const uint8_t, 3> input_image,
+        dali::kernels::TensorView<dali::kernels::StorageCPU, float, 3> output_image,
+        dali::kernels::TensorView<dali::kernels::StorageCPU, const float, 3> external_hints) {
+  auto ptr = output_image.data;
+  ptr[0] = kStubValue;
```
Comment (jantonguirao) [inline dali/aux/optical_flow/optical_flow_stub.h]: what is kStubValue?

File: dali/aux/optical_flow/optical_flow_adapter_test.cc (hunk 6 above)
Comment (szalpal) [reply dali/aux/optical_flow/optical_flow_adapter_test.cc]: Done

File: dali/aux/optical_flow/optical_flow_stub.h (hunk 7 above)
Comment (szalpal) [reply dali/aux/optical_flow/optical_flow_stub.h]: Done

File: dali/aux/optical_flow/optical_flow_stub.h (hunk 8 above)
Comment (szalpal) [reply dali/aux/optical_flow/optical_flow_stub.h]: Updated description of `kStubValue`

File: dali/aux/optical_flow/optical_flow_adapter_test.cc (hunk 5 above)
Comment (szalpal) [reply dali/aux/optical_flow/optical_flow_adapter_test.cc]: Done

File: dali/aux/optical_flow/optical_flow_adapter_test.cc (hunk 6 above)
Comment (jantonguirao) [reply dali/aux/optical_flow/optical_flow_adapter_test.cc]: Not done, it is still leaking

File: dali/aux/optical_flow/optical_flow_adapter.h [hunk 9]
```diff
@@ -16,14 +16,23 @@
 #define DALI_AUX_OPTICAL_FLOW_OPTICAL_FLOW_ADAPTER_H_
 
 #include <dali/pipeline/data/backend.h>
-#include "dali/kernels/tensor_view.h"
+#include <dali/kernels/backend_tags.h>
```
Comment (jantonguirao) [inline dali/aux/optical_flow/optical_flow_adapter.h]: should be `#include ""`

Comment (szalpal) [reply dali/aux/optical_flow/optical_flow_adapter.h]: Done

File: dali/aux/optical_flow/optical_flow_adapter_test.cc (hunk 6 above)
Comment (szalpal) [reply dali/aux/optical_flow/optical_flow_adapter_test.cc]: Now done :)

**Pull request merged**